query-planner-js/src/Scope.ts

```diff
@@ -0,0 +1,259 @@
+   * @returns the newly created scope.
+   */
+  static create(context: QueryPlanningContext, parentType: GraphQLCompositeType) {
+    return new Scope(context, parentType, undefined, undefined);
+  }
+
+  /**
+   * Refines this scope by (maybe) adding the provide type (with the optional provided directives).
+   *
+   * @param type - the new type to refine this scope with.
+   * @param directives - optional directives for `type` condition.
+   * @returns a scope corresponding to restrictive this scope with the `type`. This may return
+   * this scope (be a no-op) if `directives` is undefined/null/empty and `type` is a super type of
+   * one of the existing type in this scope. Otherwise (if either `directives` is non empty or
+   * `type` genuinely restricts the existing types in this scope), a new scope with `type`
+   * as first link in the scope chain.
+   */
+  refine(type: GraphQLCompositeType, directives?: ReadonlyArray<DirectiveNode>) : Scope {
+    // Always treat the absence of directives as "undefined" to make is simpler.
+    if (directives && directives.length == 0) {
+      directives = undefined;
+    }
+    // If we have directives, we always want to preserve the condition so as to preserve that directive.
+    if (directives) {
+      return new Scope(this.context, type, directives, this);
+    }
+    // Scope always preserve the immediate parent type, but if the new type is already the parent one
+    // and we have not directives to preserve, simply return the existing scope.
+    if (!directives && type === this.parentType) {
+      return this;
+    }
+    return new Scope(this.context, type, directives, Scope.pruneRefinedTypes(this, type));
+  }
+
+  private static pruneRefinedTypes(
+    toPrune: Scope | undefined,
+    refiningType: GraphQLCompositeType
+  ) : Scope | undefined {
+    if (!toPrune) {
+      return undefined;
+    }
+    if (!toPrune.directives && isTypeSubTypeOf(toPrune.context.schema, refiningType, toPrune.parentType)) {
+      // The newly added type is a subtype of the current "link", and the current link has no directives,
+      // so it's not useful anymore. Skip it, and check if we can prune further.
+      return Scope.pruneRefinedTypes(toPrune.enclosing, refiningType);
+    }
+    return new Scope(
+      toPrune.context,
+      toPrune.parentType,
+      toPrune.directives,
+      Scope.pruneRefinedTypes(toPrune.enclosing, refiningType)
+    );
+  }
+
+  /**
+   *  Whether this scope is restricting the possible runtime types of the provided type.
+   */
+  isStrictlyRefining(type: GraphQLCompositeType) : boolean {
+    // This scope will refine the provided type, unless that provided type is a subtype of all
+    // the type in the chain.
+    let scope: Scope | undefined = this;
+    while (scope) {
+      if (scope.parentType !== type && isTypeSubTypeOf(this.context.schema, scope.parentType, type)) {
+        return true;
+      }
+      scope = scope.enclosing;
```

Comment:
This falls more under the category of "might be interesting to play with", but since the pattern of iterating over a scope chain is pretty common, that might be a nice use for a generator:

```
  *wholeChain(): Generator<Scope> {
    yield this;
    if (this.enclosing) {
      yield *this.enclosing.wholeChain();
    }
  }
```